rasterio/_features.pyx

```diff
@@ -74,7 +72,7 @@ def _shapes(image, mask, connectivity, transform):
         hband = mem_ds.band
     elif isinstance(image, tuple):
         rdr = image.ds
-        hband = rdr.band(image.bidx)
+        hband = (<_io.DatasetReaderBase?>rdr).band(image.bidx)
```

Comment:
@perrygeo now that we're deriving a Python class from the base class, this is the proper thing to do for our C code.